File: clustering/clustering_complete_code.R


## Clustering Complete Code

## Importing the dataset
df = read.csv('Mall_Customers.csv')
df = df[4:5]

## Fitting the Clusters 

#-------------------------------------------------------# 

## K-Means

## Finding optimal K (no of cluster) using elbow method (reduce WCSS - Within Cluster Sum of Square) 
title = 'K-Means'
library(class)
set.seed(123)
wcss = vector()
for (i in 1:10)
  wcss[i] = sum(kmeans(df,i)$withinss)

plot(c(1:10), wcss, 
     type='b',
     main= 'Elbow Method',
     xlab = 'No of Clusters',
     ylab = 'WCSS')

set.seed(1234)
clust = kmeans(df, 5)
y_pred = clust$cluster


#------------------------- OR ----------------------------#

## Hierarchical Clustering 

## Finding optimal no of cluster using dendrogram method (reduce the variance within clusters)

title = 'Hierarchical Clustering'
dendrogram = hclust(d = dist(df, method = 'euclidean'), method = 'ward.D')
plot(dendrogram,
     main = 'Dendrogram',
     xlab = 'Customers',
     ylab = 'Euclidean Distances')


hc = hclust(d = dist(df, method = 'euclidean'), method = 'ward.D')
y_pred = cutree(hc, 5)

#-------------------------------------------------------#

## Visualising the results for clusters,

## K-Means Or Hierarchichal Clustering

library(cluster)
clusplot(df,
         y_pred,
         lines=0,
         color=T,
         labels=4,
         plotchar = F,
         span = T,
         main = paste('Cluster Plot (' , title , ')'),
         xlab = 'Annual Income (k$)',
         ylab = 'Spending Score (1-100)')

File: clustering/clustering_complete_code.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jan 29 19:12:07 2020

@author: Amit Anchalia

@Description: Clustering Complete Code
"""

## Importing the library 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Importing the dataset
df = pd.read_csv('Mall_Customers.csv')
X = df.iloc[:, 3:5]

## Fitting the Clusters

#-------------------------------------------------------#

## K-Mean

## Finding optimal K (no of cluster) using elbow method (reduce WCSS - Within Cluster Sum of Square) 
from sklearn.cluster import KMeans
wcss = []
for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, init='k-means++', random_state=0)
    kmeans.fit(X)
    wcss.append(kmeans.inertia_)
    
plt.plot(range(1,11), wcss)
plt.title('Elbow Method')
plt.xlabel('No of Clusters')
plt.ylabel('WCSS')
plt.show

from sklearn.cluster import KMeans
title = 'K-Means'
cluster = KMeans(n_clusters = 5, init='k-means++', random_state=0)
y_pred  = cluster.fit_predict(X)

#------------------------- OR ----------------------------#

## Hierarchical Clustering

## Finding optimal no of cluster using dendrogram method (reduce the variance within clusters)
import scipy.cluster.hierarchy as sch
dendrogram = sch.dendrogram(sch.linkage(X, method='ward'))
plt.title('Dendrogram')
plt.xlabel('Customers')
plt.ylabel('Euclidean distances')
plt.show()

from sklearn.cluster import AgglomerativeClustering
title = 'Hierarchical Clustering'
cluster = AgglomerativeClustering(n_clusters=5, affinity='euclidean', linkage='ward')
y_pred  = cluster.fit_predict(X)

## Visualizing the clusters

#-------------------------------------------------------#
title_ = title + ' (Clusters of Customers)' 
colors = ['red', 'green', 'blue', 'cyan', 'orange']
for i in range(0,5):
    label_ = 'cluster' + str(i)
    plt.scatter(X.iloc[y_pred == i, 0], X.iloc[y_pred == i, 1], s=10, c=colors[i], label=label_)

plt.title(title_)
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend()
plt.show()